docs/source/explanations/dataclass.md

```diff
@@ -0,0 +1,455 @@
+4. Provide a [**`default_factory`
+   function**](https://docs.python.org/3/library/dataclasses.html#default-factory-functions)
+   to generate the default value for a field using a function (this is important
+   for [mutable default
+   values](https://docs.python.org/3/library/dataclasses.html#mutable-default-values)
+   such as an empty list.)
+    ```python
+    from dataclasses import dataclass, field
+
+    @dataclass
+    class D:
+        x: list = field(default_factory=list)
+    ```
+5. Dataclasses *can* be [used as a base class for
+   **inheritance**](https://docs.python.org/3/library/dataclasses.html#inheritance),
+   but know that there are some tricky scenarios; specifically: the subclass
+   cannot have any non-defaut arguments if the parent class defines any fields
+   with defaults. (see [this stack-overflow
+   post](https://stackoverflow.com/questions/51575931/class-inheritance-in-python-3-7-dataclasses)
+   for details)
+
+In some cases, you may also need to modify the behavior of the full dataclass, or of individual fields in the dataclass using the `init`, `repr`, `order`, `unsafe_hash`, or `frozen` parameters as well, so it's worth [reviewing their function](https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass).
+
+## napari's `evented_dataclass`
+
+The `@evented_dataclass` decorator adds two more parameters to the standard library `dataclass`: `events` and `properties`.  **By default, they are both `True`.**
+
+### `@evented_dataclass(events=True)`
+
+When `events` is `True`, an event will be emitted each time an attribute is modified, as described in the introduction.  For example...
+
+```python
+@evented_dataclass(events=True, properties=False)
+class A:
+    x: int = 1
+```
+
+Is *roughly* equivalent to:
+
+```python
+class A:
+    def __init__(self, x: int = 1):
+        self.x = x
+        self.events = EmitterGroup(source=self, x=None)
+
+    def __setattr__(self, name, value):
+        before = getattr(self, name) # get current value before changing
+        object.__setattr__(self, name, value)  # set new value
+
+        # if custom set method `_on_<name>_set` exists, call it
+        if custom_setter := getattr(self, f'_on_{name}_set', None):
+            # if it returns a truthy value, return without emitting event
+            if custom_setter(getattr(self, name)):
+                return
+
+        # finally, if the value has changed, emit the appropriate event
+        if before != getattr(self, name):
+            event_emitter = getattr(self.events, name)  # get emitter
+            event_emitter(value=after)  # emit event
+```
+
+### `@evented_dataclass(properties=True)`
+
+When `properties` is `True`, attributes are turned in to `property` descriptors that set a private attribute, with the possibility of setting custom getter/setter functions using the special `Property` type.  (In this case, it is possible that the internal/private type differs from the external public value)
```

Comment:
I find this such a useful part of the evented_dataclass that I might mention it in the intro about events too. While the original motivation wasn't to add this syntax I find it very clean and a big benefit

Reply:
I agree the `Property` type syntax is super nice.

I think it would be nice to add a sentence giving some guidance on using the `Property` vs. `_on_{prop}_set()` methods. As I understand things so far, it seems like the `Property` type is useful for simple type coercion and the `_on_{prop}_set()` methods are better when something more complex is required such as modifying other private attributes. However, I am curious as to what your thoughts are, @tlambert03 and @sofroniewn !